File: Go/Algorithmic_Blueprints/Sorting_and_Order_Statistics/insertion_sort/insertion_sort.go
package main

import "fmt"

// InsertionSort sorts an array of integers using the insertion sort algorithm.
func InsertionSort(arr []int) {
	n := len(arr)
	for i := 1; i < n; i++ {
		key := arr[i]
		j := i - 1
		// Move elements of arr[0..i-1] that are greater than key
		// to one position ahead of their current position
		for j >= 0 && arr[j] > key {
			arr[j+1] = arr[j]
			j = j - 1
		}
		arr[j+1] = key
	}
}

func main() {
	sample := []int{64, 34, 25, 12, 22, 11, 90}
	fmt.Println("Original array:", sample)
	InsertionSort(sample)
	fmt.Println("Sorted array:  ", sample)
}

File: python/algorithms/sorting/insertion_sort.py

def insertion_sort(arr):
    """
    Sorts a list in ascending order using the Insertion Sort algorithm.

    Args:
        arr (list): A list of comparable elements.

    Returns:
        list: The sorted list.
    """
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1

        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1

        arr[j + 1] = key

    return arr


def main():
    try:
        data = [2, 1, 4, 5, 3]

        sorted_data = insertion_sort(data.copy())
        print("Insertion Sort:", sorted_data)

    except Exception as e:
        print("Exception:", e)


if __name__ == "__main__":
    main()
